rasa/nlu/training_data/formats/markdown.py

```diff
@@ -211,23 +211,31 @@ def dumps(self, training_data: "TrainingData") -> Text:
     def _generate_training_examples_md(self, training_data: "TrainingData") -> Text:
         """Generates markdown training examples."""
 
-        # Sort training data by intent
-        training_examples = sorted(
-            [e.as_dict_nlu() for e in training_data.training_examples],
-            key=lambda k: k[MESSAGE_INTENT_ATTRIBUTE],
-        )
+        from collections import defaultdict
+
+        training_examples = defaultdict(list)
+
+        # Sort by intent while keeping basic intent order
+        for example in [e.as_dict_nlu() for e in training_data.training_examples]:
+            intent = example[MESSAGE_INTENT_ATTRIBUTE]
+            training_examples[intent].append(example)
+
         md = ""
-        for i, example in enumerate(training_examples):
-            intent = training_examples[i - 1][MESSAGE_INTENT_ATTRIBUTE]
-            if i == 0 or intent != example[MESSAGE_INTENT_ATTRIBUTE]:
-                md += self._generate_section_header_md(
-                    INTENT,
-                    example[MESSAGE_INTENT_ATTRIBUTE],
-                    example.get(MESSAGE_RESPONSE_KEY_ATTRIBUTE, None),
-                    i != 0,
-                )
 
-            md += self._generate_item_md(self._generate_message_md(example))
+        for intent, examples in training_examples.items():
+            for i, example in enumerate(examples):
+                did_intent_change = i == 0
+                is_first_line = md == ""
+
+                if did_intent_change:
```

Comment:
I think it makes sense to move this out of the inner loop:
```python
 for intent, examples in training_examples.items():
            if examples:
              md += self._generate_section_header_md(
                        INTENT,
                        intent,
                        example.get(MESSAGE_RESPONSE_KEY_ATTRIBUTE, None),
                        not is_first_line,
                    )
            for i, example in enumerate(examples):

Reply:
tbh. the newline (`not is_first_line`) should probably not handled in that `_generate_section_header_md ` function but it should rather work smth like this:
```
sections = []
for intent, examples in training_examples.items():
  # create markdown for section ...
  sections.append(section)

return "\n".join(sections)

Reply:
Can do 👍 I didn't want to mess with it too much for this quick change.